internal/db/bootstrap_logs.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2025 Canonical.
+
+package db
+
+import (
+	"context"
+
+	"github.com/google/uuid"
+
+	"github.com/canonical/jimm/v3/internal/dbmodel"
+	"github.com/canonical/jimm/v3/internal/errors"
+	"github.com/canonical/jimm/v3/internal/servermon"
+)
+
+func (d *Database) AddBootstrapLog(ctx context.Context, jobId uuid.UUID, logLine string) (err error) {
+	const op = errors.Op("db.AddBootstrapLog")
+
+	if err := d.ready(); err != nil {
+		return errors.E(op, err)
+	}
+
+	// Get the current line number for this bootstrap job.
+	var currentLineNumber int
+	err = d.DB.WithContext(ctx).
+		Model(&dbmodel.BootstrapLog{}).
+		Where("job_id = ?", jobId).
+		Select("COALESCE(MAX(line_number), 0)").
+		Scan(&currentLineNumber).Error
+	if err != nil {
+		return errors.E(op, "failed to get current line number", err)
+	}
+
+	nextLineNumber := currentLineNumber + 1
+
+	log, err := dbmodel.NewBootstrapLog(jobId, nextLineNumber, logLine)
+	if err != nil {
+		return errors.E(op, "failed to construct bootstrap log", err)
+	}
+
+	if err := d.DB.WithContext(ctx).Create(log).Error; err != nil {
+		return errors.E(op, dbError(err))
+	}
+
+	return nil
+}
+
+// QueryBootstrapLog queries for bootstrap logs based on the jobId and offset.
+// It returns an error if the job doesn't exist, there aren't any logs or the offset
+// is greater than the amount of logs,
+func (d *Database) QueryBootstrapLog(ctx context.Context, jobId uuid.UUID, offset int) (loggies []string, err error) {
+	const op = errors.Op("db.QueryBootstrapLog")
+
+	if err := d.ready(); err != nil {
+		return loggies, errors.E(op, err)
+	}
+
+	durationObserver := servermon.DurationObserver(servermon.DBQueryDurationHistogram, string(op))
+	defer durationObserver()
+	defer servermon.ErrorCounter(servermon.DBQueryErrorCount, &err, string(op))
+
+	// Make sure job exists, if it doesn't, there's no point running the query
+	if err := d.DB.WithContext(ctx).First(&dbmodel.JobTrackerEntry{JobID: jobId}, "job_id = ?", jobId).Error; err != nil {
+		return loggies, errors.E(op, "job not found", dbError(err))
+	}
+
+	query := d.DB.WithContext(ctx).
+		Model(&dbmodel.BootstrapLog{}).
+		Where("job_id = ?", jobId).
+		Order("line_number ASC")
+
+	var count int64
+	if err := query.Count(&count).Error; err != nil {
+		return loggies, errors.E(op, dbError(err))
+	}
+
+	if count == 0 {
+		return loggies, errors.E(op, "no logs")
+	}
+
+	// Validate the offset isn't greater than the amount of actual logs
+	if int64(offset) >= count {
+		return loggies, errors.E(op, "offset cannot be greater than or equal to the amount of logs")
+	}
```

Comment:
Also, I think this check should go before `if count == 0` otherwise you get different behaviour when offset=N and count == 0 vs when offset=N and count != 0

Reply:
I found this SO discussion what the best recommend way to handle an offset larger than the amount of items.
See [here](https://stackoverflow.com/questions/11402156/rest-status-code-204-on-paginated-result) and [here](https://stackoverflow.com/questions/32081423/rest-api-pagination-how-to-handle-offset-exceeding-total-count).
The suggestion is to return an HTTP 204 (No Content). Since we're using websockets I'm okay if you return the error or an empty set of results to the client. 
My only suggestion then would be to make the error the same as when `count == 0`.

Reply:
Added a code not found to the error.